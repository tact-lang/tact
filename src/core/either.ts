// left$0 {X:Type} {Y:Type} value:X = Either X Y;
// right$1 {X:Type} {Y:Type} value:Y = Either X Y;

import { uint } from "@/core/numeric";
import type { Tlb } from "@/core/tlb";

export type Either<L, R> = Left<L> | Right<R>
export type Left<L> = { readonly $: 'left', readonly value: L }
export const Left = <L>(value: L): Left<L> => Object.freeze({ $: 'left', value });
export const isLeft = <L>(m: Either<L, unknown>): m is Left<L> => m.$ === 'left';
export type Right<R> = { readonly $: 'right', readonly value: R }
export const Right = <R>(value: R): Right<R> => Object.freeze({ $: 'right', value });
export const isRight = <R>(m: Either<unknown, R>): m is Right<R> => m.$ === 'right';

export const either = <L, R>(left: Tlb<L>, right: Tlb<R>): Tlb<Either<L, R>> => ({
    store: (t, b, p) => {
        if (t.$ === 'left') {
            uint(1).store(0, b, [...p, '$']);
            left.store(t.value, b, [...p, 'value']);
        } else {
            uint(1).store(1, b, [...p, '$']);
            right.store(t.value, b, [...p, 'value']);
        }
    },
    load: (s, p) => {
        const type = uint(1).load(s, [...p, '$']);
        if (type === 0) {
            return Left(left.load(s, [...p, 'value']));
        } else {
            return Right(right.load(s, [...p, 'value']));
        }
    },
});
