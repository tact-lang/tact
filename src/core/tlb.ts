import { entries } from '@/utils/tricks';
import { type Slice, Builder } from '@ton/core'

export type Path = string[]

export class TlbError extends Error {
    constructor(
        public error: unknown,
        public context: Path
    ) {
        super();
    }
}

/**
 * TL-B type
 */
export interface Tlb<T> {
    /**
     * Store `value` into `builder` in context `path`
     */
    store: (value: T, builder: Builder, path: Path) => void;
    /**
     * Load value from `slice` in context `path`
     */
    load: (slice: Slice, path: Path) => T;
}

export type GetTlb<T> = T extends Tlb<infer U> ? U : never

/**
 * Internal wrapper for primitive types
 * Catches errors if load/store failed
 */
export const _primitive = <T>(child: Tlb<T>): Tlb<T> => ({
    store: (t, b, p) => {
        try {
            child.store(t, b, p)
        } catch (e) {
            throw new TlbError(e, p);
        }
    },
    load: (s, p) => {
        try {
            return child.load(s, p);
        } catch (e) {
            throw new TlbError(e, p);
        }
    },
});

/**
 * Non-stored value
 * Always returns `value` when loaded
 * Doesn't have any binary representation
 */
export const pure = <T>(value: T): Tlb<T> => ({
    store: (_t, _b) => {},
    load: (_s) => value,
});

/**
 * Ref to another type
 * Equivalent to `^child` in TL-B
 */
export const ref = <T>(child: Tlb<T>): Tlb<T> => ({
    store: (t, b, p) => {
        const b2 = new Builder();
        const refId = `refs[${b.refs}]`;
        const path = [...p, refId];
        child.store(t, b2, path);
        try {
            b.storeRef(b2.endCell());
        } catch (e) {
            throw new TlbError(e, path);
        }
    },
    load: (s, p) => {
        const refId = `refs[${s.offsetRefs}]`;
        const path = [...p, refId];
        try {
            return child.load(s.loadRef().asSlice(), path);
        } catch (e) {
            throw new TlbError(e, path);
        }
    },
});

/**
 * Sequence of types
 * Equivalent to space-separated list in TL-B:
 * `foo:T bar:U`
 */
export const object = <T extends object>(
    children: { [K in keyof T]: Tlb<T[K]> }
): Tlb<T> => {
    const fields = entries(children);
    return {
        store: (t, b, p) => {
            for (const [k, child] of fields) {
                child.store(t[k], b, [...p, String(k)]);
            }
        },
        load: (s, p) => {
            const result = {} as T;
            for (const [k, child] of fields) {
                result[k] = child.load(s, [...p, String(k)]);
            }
            return result;
        },
    };
};

/**
 * Exact value of type `child`
 * Equivalent to `#1234` in TL-B
 */
export const literal = <const T>(child: Tlb<T>) =>
    <U extends T>(value: U): Tlb<U> => ({
    store: (t, b, p) => {
        if (t !== value) {
            throw new TlbError(new Error('Invalid value'), p);
        }
        child.store(value, b, p);
    },
    load: (s, p) => {
        const readValue = child.load(s, p);
        if (value !== readValue) {
            throw new TlbError(new Error('Invalid value'), p);
        }
        return value;
    },
});