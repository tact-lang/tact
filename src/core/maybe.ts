// nothing$0 {X:Type} = Maybe X;
// just$1 {X:Type} value:X = Maybe X;

import { TlbError, type Path, type Tlb } from "@/core/tlb";
import type { Builder, Slice } from "@ton/core";

export type Maybe<T> = Nothing | Just<T>
export type Nothing = { readonly $: 'nothing' }
export const Nothing: Nothing = Object.freeze({ $: 'nothing' })
export const isNothing = <T>(m: Maybe<T>): m is Nothing => m.$ === 'nothing';
export type Just<T> = { readonly $: 'just', readonly value: T }
export const Just = <T>(value: T): Just<T> => Object.freeze({ $: 'just', value });
export const isJust = <T>(m: Maybe<T>): m is Just<T> => m.$ === 'just';

export const maybe = <T>(type: Tlb<T>): Tlb<Maybe<T>> => {
    const store = (t: Maybe<T>, b: Builder, p: Path) => {
        if (t.$ === 'nothing') {
            b.storeBit(0);
            return;
        }
        try {
            b.storeBit(1);
        } catch (e) {
            throw new TlbError(e, [...p, '$']);
        }
        type.store(t.value, b, [...p, 'value']);
    };
    const load = (s: Slice, p: Path) => {
        const bit = (() => {
            try {
                return s.loadBit();
            } catch (e) {
                throw new TlbError(e, [...p, '$']);
            }
        })();
        if (bit) {
            return Just(type.load(s, [...p, 'value']));
        } else {
            return Nothing;
        }
    };
    return { store, load };
};