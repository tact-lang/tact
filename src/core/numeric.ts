import { _primitive, type Tlb } from "@/core/tlb";

export const bit: Tlb<boolean> = _primitive({
    store: (t, b) => b.storeBit(t),
    load: (s) => s.loadBit(),
});

export const int = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeInt(t, bits),
    load: (s) => s.loadInt(bits),
});

export const uint = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeUint(t, bits),
    load: (s) => s.loadUint(bits),
});

export const intVar = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeVarInt(t, bits),
    load: (s) => s.loadVarInt(bits),
});

export const uintVar = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeVarUint(t, bits),
    load: (s) => s.loadVarUint(bits),
});

export const coins = uintVar(16);

export const intBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeInt(t, bits),
    load: (s) => s.loadIntBig(bits),
});

export const uintBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeUint(t, bits),
    load: (s) => s.loadUintBig(bits),
});

export const intVarBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeVarInt(t, bits),
    load: (s) => s.loadVarIntBig(bits),
});

export const uintVarBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeVarUint(t, bits),
    load: (s) => s.loadVarUintBig(bits),
});