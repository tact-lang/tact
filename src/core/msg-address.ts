// addr_none$00 = MsgAddressExt;
// addr_extern$01 len:(## 9) external_address:(bits len) = MsgAddressExt;
// anycast_info$_ depth:(#<= 30) { depth >= 1 }
//    rewrite_pfx:(bits depth) = Anycast;
// addr_std$10 anycast:(Maybe Anycast) workchain_id:int8 address:bits256  = MsgAddressInt;
// addr_var$11 anycast:(Maybe Anycast) addr_len:(## 9) workchain_id:int32 address:(bits addr_len) = MsgAddressInt;
// _ _:MsgAddressInt = MsgAddress;
// _ _:MsgAddressExt = MsgAddress;

import { int, uint, uintBig } from "@/core/numeric";
import type { Path, Tlb } from "@/core/tlb";
import type { Builder, Slice } from "@ton/core";

export type MsgAddress =
    | AddrNone
    | AddrExtern
    | AddrStd
    | AddrVar
export type AddrNone = {
    readonly $: 'AddrNone'
}
export const AddrNone: AddrNone = Object.freeze({ $: "AddrNone" });
export const isAddrNone = (a: MsgAddress): a is AddrNone => a.$ === 'AddrNone';
export type AddrExtern = {
    readonly $: 'AddrExtern'
    readonly len: number
    readonly address: bigint; // bits len
}
export const AddrExtern = (len: number, address: bigint): AddrExtern => Object.freeze({
    $: 'AddrExtern', address, len,
})
export const isAddrExtern = (a: MsgAddress): a is AddrExtern => a.$ === 'AddrExtern';
export type AddrStd = {
    readonly $: 'AddrStd'
    readonly workchainId: number;
    readonly address: bigint;
}
export const AddrStd = (workchainId: number, address: bigint): AddrStd => Object.freeze({
    $: 'AddrStd', address, workchainId,
})
export const isAddrStd = (a: MsgAddress): a is AddrStd => a.$ === 'AddrStd';
export type AddrVar = {
    readonly $: 'AddrVar'
    readonly len: number // ## 9
    readonly workchainId: number // int32
    readonly address: bigint // bits addr_len
}
export const AddrVar = (len: number, workchainId: number, address: bigint): AddrVar => Object.freeze({
    $: "AddrVar", address, len, workchainId,
});
export const isAddrVar = (a: MsgAddress): a is AddrVar => a.$ === 'AddrVar';

const loadAddress = (s: Slice, p: Path): MsgAddress => {
    const type = uint(2).load(s, [...p, '$']);
    if (type === 0) {
        // addr_none$00
        return AddrNone;
    } else if (type === 1) {
        // addr_extern$01
        // len:(## 9)
        const len = uint(9).load(s, [...p, 'len']);
        // external_address:(bits len)
        const addr = uintBig(len).load(s, [...p, 'address']);
        return AddrExtern(len, addr);
    } else if (type === 2) {
        // addr_std$10
        // anycast:(Maybe Anycast)
        if (uint(1).load(s, [...p, 'anycast']) !== 0) {
            throw Error('Anycast not supported');
        }
        // workchain_id:int8
        const wc = int(8).load(s, [...p, 'workchainId']);
        // address:bits256
        const addr = uintBig(256).load(s, [...p, 'address']);
        return AddrStd(wc, addr);
    } else if (type === 3) {
        // addr_var$11
        // anycast:(Maybe Anycast)
        if (uint(1).load(s, [...p, 'anycast']) !== 0) {
            throw Error('Anycast not supported');
        }
        // addr_len:(## 9)
        const len = uint(9).load(s, [...p, 'len']);
        // workchain_id:int32
        const wc = int(32).load(s, [...p, 'workchainId']);
        // address:(bits addr_len)
        const addr = uintBig(len).load(s, [...p, 'address']);
        return AddrVar(len, wc, addr);
    } else {
        throw Error('Unreachable');
    }
};

const storeAddress = (t: MsgAddress, b: Builder, p: Path) => {
    switch (t.$) {
        case 'AddrNone': {
            // addr_none$00
            uint(2).store(0, b, [...p, '$']);
            return;
        }
        case 'AddrExtern': {
            // addr_extern$01
            uint(2).store(1, b, [...p, '$']);
            // len:(## 9)
            uint(9).store(t.len, b, [...p, 'len']);
            // external_address:(bits len)
            uintBig(t.len).store(t.address, b, [...p, 'address']);
            return;
        }
        case 'AddrStd': {
            // addr_std$10
            uint(2).store(2, b, [...p, '$']);
            // anycast:(Maybe Anycast)
            uint(1).store(0, b, [...p, 'anycast']);
            // workchain_id:int8
            int(8).store(t.workchainId, b, [...p, 'workchainId']);
            // address:bits256
            uintBig(256).store(t.address, b, [...p, 'address']);
            return;
        }
        case 'AddrVar': {
            // addr_var$11
            uint(2).store(3, b, [...p, '$']);
            // anycast:(Maybe Anycast)
            uint(1).store(0, b, [...p, 'anycast']);
            // addr_len:(## 9)
            uint(9).store(t.len, b, [...p, 'len']);
            // workchain_id:int32
            int(32).store(t.workchainId, b, [...p, 'workchainId']);
            // address:(bits addr_len)
            uintBig(t.len).store(t.address, b, [...p, 'address']);
        }
    }
};

export const msgAddress: Tlb<MsgAddress> = ({
    store: storeAddress,
    load: loadAddress,
});