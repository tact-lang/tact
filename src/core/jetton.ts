// InternalMsgBody
//     transfer
//         #0f8a7ea5
//         query_id:uint64
//         amount:(VarUInteger 16)
//         destination:MsgAddress
//         response_destination:MsgAddress
//         custom_payload:(Maybe ^Cell)
//         forward_ton_amount:(VarUInteger 16)
//         forward_payload:(Either Cell ^Cell)
//     transfer_notification
//         #7362d09c
//         query_id:uint64
//         amount:(VarUInteger 16)
//         sender:MsgAddress
//         forward_payload:(Either Cell ^Cell)
//     excesses
//         #d53276db
//         query_id:uint64
//     burn
//         #595f07bc
//         query_id:uint64
//         amount:(VarUInteger 16)
//         response_destination:MsgAddress
//         custom_payload:(Maybe ^Cell)
//     internal_transfer
//         #178d4519
//         query_id:uint64
//         amount:(VarUInteger 16)
//         from:MsgAddress
//         response_address:MsgAddress
//         forward_ton_amount:(VarUInteger 16)
//         forward_payload:(Either Cell ^Cell)
//     burn_notification
//         #7bdd97de
//         query_id:uint64
//         amount:(VarUInteger 16)
//         sender:MsgAddress
//         response_destination:MsgAddress

import { either } from "@/core/either";
import { maybe } from "@/core/maybe";
import { msgAddress } from "@/core/msg-address";
import { uint, uintVar } from "@/core/numeric";
import { type GetTlb, literal, object, ref, type Tlb } from "@/core/tlb";

export const transfer = <Custom, Forward>(
    custom: Tlb<Custom>,
    forward: Tlb<Forward>,
) => {
    return object({
        $: literal(uint(32))(0x0f8a7ea5),
        queryId: uint(64),
        amount: uintVar(16),
        destination: msgAddress,
        responseDestination: msgAddress,
        customPayload: maybe(ref(custom)),
        forwardTonAmount: uintVar(16),
        forwardPayload: either(forward, ref(forward)),
    });
};

export type Transfer<T, U> = GetTlb<ReturnType<typeof transfer<T, U>>>
