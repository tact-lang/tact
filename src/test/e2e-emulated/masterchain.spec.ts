import { toNano } from "@ton/core";
import { Blockchain, SandboxContract, TreasuryContract } from "@ton/sandbox";
import { MasterchainTester } from "./contracts/output/masterchain_MasterchainTester";
import "@ton/test-utils";

describe("masterchain", () => {
    let blockchain: Blockchain;
    let treasure: SandboxContract<TreasuryContract>;

    beforeEach(async () => {
        blockchain = await Blockchain.create();
        blockchain.verbosity.print = false;
        treasure = await blockchain.treasury("treasure");
    });

    //
    // Deployment and simple message receiving
    //

    it("should deploy to workchain", async () => {
        const contract = blockchain.openContract(
            await MasterchainTester.fromInit(),
        );

        const deployResult = await contract.send(
            treasure.getSender(),
            { value: toNano("10") },
            "DeployToWorkchain",
        );

        expect(deployResult.transactions).toHaveTransaction({
            from: treasure.address,
            to: contract.address,
            success: true,
            deploy: true,
        });
    });

    it("should not deploy to workchain from masterchain", async () => {
        const treasure = await blockchain.treasury("treasure", {
            workchain: -1,
        });
        const contract = blockchain.openContract(
            await MasterchainTester.fromInit(),
        );

        const deployResult = await contract.send(
            treasure.getSender(),
            { value: toNano("10") },
            "DeployToWorkchain",
        );

        expect(deployResult.transactions).toHaveTransaction({
            from: treasure.address,
            to: contract.address,
            success: false,
            deploy: true,
            exitCode: 1137,
        });
    });

    it("should deploy to masterchain from masterchain", async () => {
        const treasure = await blockchain.treasury("treasure", {
            workchain: -1,
        });
        const contract = blockchain.openContract(
            await MasterchainTester.fromInit(),
        );

        const deployResult = await contract.send(
            treasure.getSender(),
            { value: toNano("10") },
            "DeployToMasterchain",
        );

        expect(deployResult.transactions).toHaveTransaction({
            from: treasure.address,
            to: contract.address,
            success: true,
            deploy: true,
        });
    });
});
