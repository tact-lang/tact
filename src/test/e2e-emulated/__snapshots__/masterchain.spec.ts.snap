// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`masterchain should deploy to the workchain 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "Deploy",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQDSuAQHFrqpJA5TD7sDh45U7-bg76NNNfmaX7AID1bdlWiE",
          "type": "internal",
          "value": "10",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 3823n,
      },
    ],
  },
]
`;

exports[`masterchain should deploy to the workchain from masterchain when masterchain enabled 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "Deploy",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQDODwXkLheXHjo3o0OlxiKH0j0uyVUq4US8ZHZkeHjcWl8D",
          "type": "internal",
          "value": "10",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 3687n,
      },
    ],
  },
]
`;

exports[`masterchain should deploy to the workchain when masterchain enabled 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "Deploy",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQDODwXkLheXHjo3o0OlxiKH0j0uyVUq4US8ZHZkeHjcWl8D",
          "type": "internal",
          "value": "10",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 3687n,
      },
    ],
  },
]
`;

exports[`masterchain should not deploy to the workchain from masterchain 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "Deploy",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQDSuAQHFrqpJA5TD7sDh45U7-bg76NNNfmaX7AID1bdlWiE",
          "type": "internal",
          "value": "10",
        },
      },
      {
        "$type": "failed",
        "errorCode": 137,
        "errorMessage": "Masterchain support is not enabled for this contract",
      },
      {
        "$type": "sent-bounced",
        "message": {
          "body": {
            "cell": "x{FFFFFFFF000000004465706C6F79}",
            "type": "cell",
          },
          "bounce": false,
          "from": "kQDSuAQHFrqpJA5TD7sDh45U7-bg76NNNfmaX7AID1bdlWiE",
          "to": "@treasure(treasure)",
          "type": "internal",
          "value": "9.989",
        },
      },
    ],
  },
]
`;
